File: src/main/kotlin/org/gameflow/pass/Passable.kt
package org.gameflow.pass

import java.util.Collection
import org.gameflow.entity.Entity
import org.gameflow.Game

/**
 * Something that can be processed by an entity pass.
 */
public trait Passable {

    fun doPass(pass: Pass, game: Game) {
        if (this is Entity) pass.apply(this, game)

        doPassOnChildren(pass, game)
    }

    fun doPassOnChildren(pass: Pass, game: Game) {}

}

File: src/main/scala/org/gameflow/core/pass/Passable.scala
package org.gameflow.core.pass

import org.gameflow.core.Game
import org.gameflow.core.entity.Entity


/**
 * Something that can be processed by an entity pass.
 */
trait Passable {

  final def doPass(pass: Pass, game: Game) {
    this match {
      case entity: Entity => pass.handleEntity(entity, game)
      case _ => // Skip
    }

    doPassOnChildren(pass, game)
  }

  def doPassOnChildren(pass: Pass, game: Game) {}

}